### PR TITLE
[WIP] Batch requests to same region across batch commands

### PR DESCRIPTION
**Pull request by tabokie**
###  What have you changed?

Batch prewrite or commit requests to same region with same epoch in batch commands. Also batch get / raw_get requests to the same region. Batch is collected from multiple batch_commands sent to one single rpc thread.

This improvement will cut down scheduler task count and batch raft commands sent to raftstore. Also, prewrite conflict can be checked with faster seek.

Hopefully this will improve bulk insert performance with less kvdb operation and raft rpc message.

###  What is the type of the changes?

- Improvement (a change which is an improvement to an existing feature)

###  How is the PR tested?

Unit test

###  Does this PR affect documentation (docs) or should it be mentioned in the release notes?

No

###  Does this PR affect `tidb-ansible`?

No

### CODE REVIEW

File: src/server/service/kv.rs [hunk 1]
```diff
@@ -53,6 +53,91 @@ pub struct Service<T: RaftStoreRouter + 'static, E: Engine> {
...
+                options.txn_size = req.get_txn_size();                
+                let cmd = Command::Prewrite {
+                    ctx: req.take_context(),
+                    mutations,
+                    primary: req.take_primary_lock(),
+                    start_ts: req.get_start_version(),
+                    options,
+                };
+                match self.router.get(&ver) {
+                    Some(&idx) => {
```
Comment (zhangjinpeng1987) [inline src/server/service/kv.rs]: If two requests try to modify the same key, they can't combine in the same mini-batch.

File: src/storage/txn/scheduler.rs [hunk 2]
```diff
@@ -499,6 +562,15 @@ impl<E: Engine> MsgScheduler for Scheduler<E> {
...
+        match task {
+            Msg::ReadFinished { cid, tag, pr } => self.on_batch_read_finished(cid, req, pr, tag),
+            Msg::WriteFinished { cid, tag, pr, result } => self.on_batch_write_finished(cid, req, pr, result, tag),
+            Msg::FinishedWithErr { cid, err, .. } => self.batch_finish_with_err(cid, req, err),
+            _ => unreachable!(),
+        }
+    }
 }
 
 fn gen_command_lock(latches: &Latches, cmd: &Command) -> Lock {
```
Comment (zhangjinpeng1987) [inline src/storage/txn/scheduler.rs]: MiniBatch is write command, must acquire latches before run.

File: src/server/service/kv.rs [hunk 3]
```diff
@@ -214,25 +270,52 @@ impl MiniBatcher {
...
+                        Box::new(move |id, v: Result<Vec<Result<(), storage::Error>>, storage::Error>| {
+                            // let v = v.map_err(Error::from);
+                            let mut resp = PrewriteResponse::default();
+                            if let Some(err) = extract_region_error(&v) {
+                                resp.set_region_error(err);
+                            } else {
+                                resp.set_errors(extract_key_errors(v).into());
+                            }
+                            let mut res = batch_commands_response::Response::default();
+                            res.cmd = Some(batch_commands_response::response::Cmd::Prewrite(resp));
```
Comment (tabokie) [inline src/server/service/kv.rs]: command mismatch here

File: src/storage/mvcc/txn.rs [hunk 4]
```diff
@@ -423,6 +426,263 @@ impl<S: Snapshot> MvccTxn<S> {
...
+        } else {
+            0
+        };
+        let mut write_iter = self
+            .reader
+            .new_write_cursor(&min_key, min_ts, &max_key, u64::max_value())
+            .unwrap();
+        let mut lock_iter = self.reader.new_lock_cursor(&min_key, &max_key).unwrap();
+        write_iter.seek(
+            &min_key.clone().append_ts(min_ts),
```
Comment (zhangjinpeng1987) [inline src/storage/mvcc/txn.rs]: should be max_ts

File: src/storage/mvcc/txn.rs [hunk 5]
```diff
@@ -423,6 +426,263 @@ impl<S: Snapshot> MvccTxn<S> {
...
+        let max_key = Key::from_encoded(upper);
+        let mut min_idx = lower_idx;
+        let mut min_ts = if let Command::Prewrite { start_ts, .. } = commands[min_idx.0] {
+            start_ts
+        } else {
+            0
+        };
+        let mut write_iter = self
+            .reader
+            .new_write_cursor(&min_key, min_ts, &max_key, u64::max_value())
```
Comment (zhangjinpeng1987) [inline src/storage/mvcc/txn.rs]: s/u64::max_value/0

File: src/storage/mvcc/txn.rs (hunk 4 above)
Comment (tabokie) [reply src/storage/mvcc/txn.rs]: I checked that `Seek` returns the min key >= seeked key. Here it should give us the first write exceeding start-ts if any. Btw in latest patch I use cursor.prev() to check data constraints on earlier writes.

Comment (tabokie) [reply src/storage/mvcc/txn.rs]: > I checked that `Seek` returns the min key >= seeked key. Here it should give us the first write exceeding start-ts if any. Btw in latest patch I use cursor.prev() to check data constraints on earlier writes.

ts is encoded in reverse order.

**Comment by tabokie** [comment]
failed integration-common-test for #
6622762: https://internal.pingcap.net/idc-jenkins/blue/organizations/jenkins/tikv_ghpr_integration_common_test/detail/tikv_ghpr_integration_common_test/3263/pipeline/

File: src/storage/mvcc/txn.rs [hunk 6]
```diff
@@ -423,6 +427,240 @@ impl<S: Snapshot> MvccTxn<S> {
...
+        let max_key = Key::from_encoded(upper);
+
+        // initialize iterator over write/lock cf
+        let mut write_iter = self
+            .reader
+            .new_write_cursor(&min_key, u64::max_value(), &max_key, 0)
+            .unwrap();
+        let mut lock_iter = self.reader.new_lock_cursor(&min_key, &max_key).unwrap();
+
+        let mut range = map.range::<Vec<u8>, _>(lower..max_key.as_encoded());
```
Comment (MyonKeminta) [inline src/storage/mvcc/txn.rs]: Is the `max_key` necessary? I think leave the right end unbounded is ok.

File: src/storage/mvcc/txn.rs [hunk 7]
```diff
@@ -423,6 +427,240 @@ impl<S: Snapshot> MvccTxn<S> {
...
+            // find first active mutation
+            let min_ts = loop {
+                if let Some(tuple) = range.next() {
+                    min_key = Key::from_encoded(tuple.0.to_vec());
+                    min_idx = tuple.1;
+                } else {
+                    break 'out;
+                }
+                if ids[min_idx.0] < u64::max_value()
+                    && if let Command::Prewrite { options, .. } = &commands[min_idx.0] {
```
[the remark this answers is not in the document]
Comment (tabokie) [reply src/storage/mvcc/txn.rs]: no, it's just I can't do a static_cast to the destined type like in C++. Any suggestion to make cleaner?

File: src/storage/mvcc/reader/reader.rs [hunk 8]
```diff
@@ -68,6 +80,64 @@ impl<S: Snapshot> MvccReader<S> {
...
+            )),
+            Some(KeyBuilder::from_slice(
+                hi.clone().append_ts(hi_ts).as_encoded().as_slice(),
+                DATA_KEY_PREFIX_LEN,
+                0,
+            )),
+            self.fill_cache,
+        )
+        .use_prefix_seek()
+        .set_prefix_same_as_start(true);
```
[the remark this answers is not in the document]
Comment (tabokie) [reply src/storage/mvcc/reader/reader.rs]: as per https://github.com/facebook/rocksdb/wiki/Prefix-Seek-API-Changes, I shouldn't have set prefix_same_as_start here. Good catch!

**Comment by mahjonp** [comment]
/bench tidb=pr/12569
```
tidb.yml:
  tikv_client:
    # Max gRPC connections that will be established with each tikv-server.
    grpc-connection-count: 4
    point-get-grpc-connection-count: 4
```

**Comment by mahjonp** [comment]
/bench tidb=pr/12569 +tpcc +sysbench
```
tidb.yml:
  tikv_client:
    # Max gRPC connections that will be established with each tikv-server.
    grpc-connection-count: 4
```

**Pull request closed without merging**